### PR TITLE
S3247: introduce C#9 patterns support

### PR DESCRIPTION
**Pull request by dmitry-golovinov-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastShouldNotBeDuplicated.CSharp9.cs [hunk 1]
```diff
@@ -31,23 +32,64 @@ public void Foo(Object x, Object y)
...
             }
 
-            if (x is Fruit f2)
+            if ((x, y) is (Fruit f2, Vegetable v2))   // Secondary
             {
-                var ff = (Fruit)f2; // FN
+                var ff2 = (Vegetable)v2;              // Noncompliant
             }
 
+            if ((x, y) is (Fruit f3, Vegetable v3))   // Noncompliant
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastShouldNotBeDuplicated.CSharp9.cs:46]: I think we should add also a few cases with `is not`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastShouldNotBeDuplicated.CSharp9.cs [hunk 2]
```diff
@@ -31,23 +32,64 @@ public void Foo(Object x, Object y)
...
+                var ff9 = (Fruit)x;
+            }
+
+            x is Fruit f0; // Error [CS0201]
+
             var message = x switch
             {
-                Fruit f3 => ((Fruit)f3).ToString(), // FN
-                Vegetable v3 => ((Vegetable)v3).ToString(), // FN
+                Fruit f10 => ((Fruit)f10).ToString(), // FN
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastShouldNotBeDuplicated.CSharp9.cs:85]: We should add also a branch with a tuple pattern inside the switch

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs [hunk 3]
```diff
@@ -26,24 +27,59 @@
...
+        {
+            var isPatternExpression = (IsPatternExpressionSyntaxWrapper)analysisContext.Node;
+            if (!(isPatternExpression.SyntaxNode.GetFirstNonParenthesizedParent() is IfStatementSyntax parentIfStatement))
+            {
+                return;
+            }
+            var isPatternLocation = isPatternExpression.SyntaxNode.GetLocation();
+            if (isPatternExpression.Pattern.SyntaxNode.IsKind(SyntaxKindEx.RecursivePattern))
+            {
+                var recursivePattern = (RecursivePatternSyntaxWrapper)isPatternExpression.Pattern.SyntaxNode;
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs]: As far as I can tell we handle only the first level of the recursive pattern. 
I think we should either support imbrication or document the FN.
e.g.
```
void M(object x, object y)
    {
        if ((x, y) is ((int a, int b), string v3))
        {
            
        }
    }
```

Comment (dmitry-golovinov-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs]: Will document it as FN.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs [hunk 4]
```diff
@@ -26,55 +27,162 @@
...
-                castTypeSymbol.TypeKind == TypeKind.Struct)
+        private static void CasePatternSwitchLabel(SyntaxNodeAnalysisContext analysisContext)
+        {
+            var casePatternSwitch = (CasePatternSwitchLabelSyntaxWrapper)analysisContext.Node;
+            if (!(casePatternSwitch.SyntaxNode.GetFirstNonParenthesizedParent().GetFirstNonParenthesizedParent() is SwitchStatementSyntax parentSwitchStatement))
             {
                 return;
             }
+            ProcessPatternExpression(analysisContext, casePatternSwitch.Pattern, parentSwitchStatement.Expression, parentSwitchStatement.Expression.GetLocation(), parentSwitchStatement);
+        }
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs:63]: Empty line between methods is missing.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastShouldNotBeDuplicated.CSharp9.cs [hunk 5]
```diff
@@ -13,41 +15,108 @@ class Program
...
         {
             var fruit = x as Fruit;
-            if (fruit is not Fruit) // FN
+            if (fruit is not Fruit) // Compliant, redundant condition, not related for the current rule
             {
             }
 
             object o;
-            switch (x)
+            switch (x)                            // Noncompliant [switch-st-0]
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CastShouldNotBeDuplicated.CSharp9.cs:23]: I am not sure the suggestion to use 'as' an null check is a valid one for a switch. So, maybe we should suggest somehow to use the u variable for example in line 32. Or just say to remove the redundant cast.

**Comment by csaba-sagi-sonarsource** [comment]
This case should be also covered:
            if (x is Fruit { Prop: 1} tuttyFrutty)
            {
                var aRealFruit = (Fruit)tuttyFrutty;
                var anotherFruit = (Fruit)x;
            }

**Comment by csaba-sagi-sonarsource** [comment]
And also:
            if (x is (Fruit or Vegetable))
            {
                var fruit = (Fruit)x;
                var vegetable = (Vegetable)x;
            }

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Please cover the cases I suggested, or at least document them as FNs.

**Comment by dmitry-golovinov-sonarsource** [comment]
> 
> 
> And also:
> if (x is (Fruit or Vegetable))
> {
> var fruit = (Fruit)x;
> var vegetable = (Vegetable)x;
> }

this case was added as a FN

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs [hunk 6]
```diff
@@ -26,55 +27,205 @@
...
+                .Where(x => x.Type.WithoutTrivia().IsEquivalentTo(castType.WithoutTrivia())
+                            && IsCastOnSameSymbol(x))
+                .Select(x => x.GetLocation()).ToList();
+
+            bool IsCastOnSameSymbol(CastExpressionSyntax castExpression) =>
+                Equals(analysisContext.SemanticModel.GetSymbolInfo(castExpression.Expression).Symbol, typeExpressionSymbol);
+        }
+
+        private static void ProcessPatternExpression(SyntaxNodeAnalysisContext analysisContext,
+            SyntaxNode isPattern,
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs:139]: For some methods you aligned the parameters under each other and on some places not. See line 156. As we still do not have a clear coding style for that I would only ask you to align them to be consistent. For whichever option you opt.

Comment (dmitry-golovinov-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CastShouldNotBeDuplicated.cs]: Done!

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**